### PR TITLE
Search: Remove unexpected sidebar widgets added on theme switch in WPCOM

### LINKED ISSUE #18556: Instant Search: standard sidebar widgets can end up in the search modal sidebar when switching themes
#### Steps to reproduce the issue

1. Set up a new WordPress site with the Twenty Twenty One theme. This theme has no sidebar area.
2. Switch to a theme that does have a sidebar of sorts, like Twenty Twenty (footer sidebar)
3. Note that the Jetpack Search sidebar now has a bunch of unwanted widgets in it, like 'Archives', 'Categories' and 'Meta'

#### What I expected

Switching themes does not add a bunch of widgets to my nice search modal.

#### What happened instead
I got some bonus sidebar widgets I didn't ask for. My guess is that WordPress looks for a widget area with 'sidebar' in the name to add the widgets to, and Twenty Twenty One doesn't create one (it just has "Footer #1" and "Footer #2").

**Screenshots**

Screen Shot 2021-01-27 at 14 16 29

### PR DESCRIPTION
**Pull request by kangzj**
Fixes #18556

#### Changes proposed in this Pull Request:
#19873 fixed the 'unwanted widgets' issue when switching theme for wp.org sites, but for wpcom the issue still exists. The reason is #19873 fixed the issue only when user trigger theme switch in admin panel, while wpcom triggers theme switch by API.

This PR moves the action to run at both front end and admin panel, and as it'll potentially affect all users while switching themes, the condition to run the fix is further restricted.

#### Jetpack product discussion
pcNPJE-96-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Apply this patch D61881 to sandbox and set up hosts to sandbox public-api
- Go to a wpcom *simple* site with Jetpack Search subscription
- Switch theme to Twenty Twenty which has two sidebar area ("Footer #1" and "Footer #2"), make sure both sidebars have some widgets
- Switch to a theme that has less sidebars, say Twenty Twenty-One which has only one sidebar ("Footer").
- Ensure that the Jetpack Search sidebar has the same widgets before the theme switch, meaning it doesn't have a bunch of unwanted widgets (set in step 3), like 'Archives', 'Categories' and 'Meta' etc.

### CODE REVIEW

File: projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php [hunk 1]
```diff
@@ -717,7 +723,11 @@ public function auto_config_woo_result_format() {
 	 * @param array $old_sidebars_widgets The sidebars_widgets option value to be saved.
 	 */
 	public function save_old_sidebars_widgets( $old_sidebars_widgets = null ) {
-		$this->old_sidebars_widgets = ! is_null( $old_sidebars_widgets ) ? $old_sidebars_widgets : wp_get_sidebars_widgets();
+		// The function should only run before _wp_sidebars_changed which migrates the sidebars.
+		// So when _wp_sidebars_changed doens't exist, we should skip the logic.
```
Comment (jsnmoon) [inline projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:727]:
```suggestion
		// So when _wp_sidebars_changed doesn't exist, we should skip the logic.
```

File: projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php [hunk 2]
```diff
@@ -730,24 +740,25 @@ public function save_old_sidebars_widgets( $old_sidebars_widgets = null ) {
 	 */
 	public function remove_wp_migrated_widgets( $sidebars_widgets ) {
 		// Hook the action only when it is a theme switch i.e. $this->old_sidebars_widgets is not empty.
-		if ( empty( $this->old_sidebars_widgets ) ) {
+		// Some extra checks to make sure the hook has minmized affection.
```
Comment (jsnmoon) [inline projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:743]:
```suggestion
		// Ensure that the hook only runs when necessary. 
```

File: projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php [hunk 3]
```diff
@@ -730,24 +740,25 @@ public function save_old_sidebars_widgets( $old_sidebars_widgets = null ) {
...
+		if (
+			empty( $this->old_sidebars_widgets )
+			|| ! is_array( $this->old_sidebars_widgets )
+			|| ! is_array( $sidebars_widgets )
+			|| ! array_key_exists( static::JETPACK_INSTANT_SEARCH_SIDEBAR, $sidebars_widgets )
+			|| ! array_key_exists( static::JETPACK_INSTANT_SEARCH_SIDEBAR, $this->old_sidebars_widgets )
+			// If the new Jetpack sidebar has already got less widgets, no need to run the code following the condition.
+			// An 'equal' would work for the condition, but we want to cover 'less than' as well, just for extra insurance.
```
Comment (jsnmoon) [inline projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:751]:
```suggestion
			// If the new Jetpack sidebar already has fewer widgets, skip execution.
			// (Uses less than comparison for defensive programming.)
```

File: projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php [hunk 4]
```diff
@@ -730,24 +740,25 @@ public function save_old_sidebars_widgets( $old_sidebars_widgets = null ) {
 	 */
 	public function remove_wp_migrated_widgets( $sidebars_widgets ) {
 		// Hook the action only when it is a theme switch i.e. $this->old_sidebars_widgets is not empty.
-		if ( empty( $this->old_sidebars_widgets ) ) {
+		// Some extra checks to make sure the hook has minmized affection.
+		if (
```
Comment (jsnmoon) [inline projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:744]: Could we add additional tests for this conditional in `test-class-jetpack-instant-search.php`?

**Review by jsnmoon — COMMENTED** [review]
Works as expected in WPCOM! Could we add additional tests for the conditional introduced in this PR? 🙂

File: projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php (hunk 4 above)
Comment (kangzj) [reply projects/plugins/jetpack/modules/search/class-jetpack-instant-search.php:744]: added in de51e15

**Comment by kangzj** [comment]
> Works as expected in WPCOM! Could we add additional tests for the conditional introduced in this PR? 🙂

@jsnmoon Added :-D

File: projects/plugins/jetpack/tests/php/modules/search/test-class-jetpack-instant-search.php [hunk 5]
```diff
@@ -64,13 +58,84 @@ public function test_remove_wp_migrated_widgets() {
...
+
+		$this->assertNull( $this->get_private_member_value( static::$instant_search, 'old_sidebars_widgets' ) );
+	}
+
+	/**
+	 * Mocked sidebars_widgets data
+	 *
+	 * @since 9.9.0
+	 */
+	private function get_old_sidebars_widgets_data() {
```
Comment (jsnmoon) [inline projects/plugins/jetpack/tests/php/modules/search/test-class-jetpack-instant-search.php:99]:
```suggestion
	private function get_old_sidebars_widgets_fixture() {
```

Clarifying that the `data` here is meant to be only used within the testing environment as a `fixture` :)

File: projects/plugins/jetpack/tests/php/modules/search/test-class-jetpack-instant-search.php [hunk 6]
```diff
@@ -64,13 +58,84 @@ public function test_remove_wp_migrated_widgets() {
...
+	 *
+	 * @since 9.9.0
+	 *
+	 * @param object $object The object to operate on.
+	 * @param string $member_name Name of the private member.
+	 *
+	 * @return mixed The value of the private member.
+	 */
+	private function get_private_member_value( $object, $member_name ) {
+		$ref = new ReflectionObject( $object );
```
Comment (jsnmoon) [inline projects/plugins/jetpack/tests/php/modules/search/test-class-jetpack-instant-search.php:120]: TIL `ReflectionObject` in PHP!

**Review by jsnmoon — DISMISSED** [review]
Looks good to go! Just a small, non-blocking nitpick on a function name.

**Review by jsnmoon — APPROVED** [review]
👍 🚢 

For what it's worth, this change was ready to ship at the time of [this review](https://github.com/Automattic/jetpack/pull/19934#pullrequestreview-671636567).

**Comment by kangzj** [comment]
> 👍 🚢 
> 
> 
> 
> For what it's worth, this change was ready to ship at the time of [this review](https://github.com/Automattic/jetpack/pull/19934#pullrequestreview-671636567).

I assumed if there were any changes then, then another review is required. In this case tho, nothing really changed except for the tests, will just ship and merge on similar conditions next time thanks 😊

**Comment by kangzj** [comment]
🚢 Shipped in r226879-wpcom

**Pull request merged**